domain/schema/model.go

```diff
@@ -401,23 +402,246 @@ ON charm_storage (charm_uuid);
 `)
 }
 
+func constraintsSchema() schema.Patch {
+	return schema.MakePatch(`
+-- constraint is a reserved word in sqlite, so we use juju_constraint instead.
+CREATE TABLE juju_constraint (
+    uuid                TEXT PRIMARY KEY,
+    arch                TEXT,
+    cpu_cores           INT,
+    cpu_power           INT,
+    mem                 INT,
+    root_disk           INT,
+    root_disk_source    TEXT,
+    instance_role       TEXT,
+    instance_type       TEXT,
+    container_type_id   INT,
+    virt_type           TEXT,
+    allocate_public_ip  BOOLEAN,
+    image_id            TEXT,
+    CONSTRAINT          fk_constraint_container_type
+        FOREIGN KEY     (container_type_id)
+        REFERENCES      container_type(id)
+);
+
+CREATE TABLE constraint_tag (
+    id                  INT PRIMARY KEY,
+    constraint_uuid     TEXT NOT NULL,
+    tag                 TEXT NOT NULL,
+    CONSTRAINT          fk_constraint_tag_constraint
+        FOREIGN KEY     (constraint_uuid)
+        REFERENCES      juju_constraint(uuid)
+);
+
+CREATE TABLE constraint_space (
+    id                  INT PRIMARY KEY,
+    constraint_uuid     TEXT NOT NULL,
+    space               TEXT NOT NULL,
+    CONSTRAINT          fk_constraint_space_constraint
+        FOREIGN KEY     (constraint_uuid)
+        REFERENCES      juju_constraint(uuid)
+);
+
+CREATE TABLE constraint_zone (
+    id                  INT PRIMARY KEY,
+    constraint_uuid     TEXT NOT NULL,
+    zone                 TEXT NOT NULL,
+    CONSTRAINT          fk_constraint_zone_constraint
+        FOREIGN KEY     (constraint_uuid)
+        REFERENCES      juju_constraint(uuid)
+);
+`)
+}
+
 func nodeSchema() schema.Patch {
 	return schema.MakePatch(`
 CREATE TABLE net_node (
     uuid TEXT PRIMARY KEY
 );
 
 CREATE TABLE machine (
+    uuid              TEXT PRIMARY KEY,
+    machine_id        TEXT NOT NULL,
+    net_node_uuid     TEXT NOT NULL,
+    life_id           INT NOT NULL,
+    base              TEXT,
+    nonce             TEXT,
+    container_type_id INT NOT NULL,
+    password_hash     TEXT NOT NULL,
+    clean             BOOLEAN,
+    force_destroyed   BOOLEAN,
+    placement         TEXT,
+    agent_started_at  DATETIME,
```

Comment:
Same, why not `agent_started_at` in `instance_data`?

Reply:
because AIUI the `instance_data` is modelling the cloud provided instance, where the `hostname` and the `agent_started_at` are both "juju machine" concerns. @wallyworld @manadart thoughts?

Reply:
agent_started_at is nothing to do with what the instance data is modelling - it's a juju concern rather than a cloud concern.
IMO hostname is intrinsically a property of the machine, like block devices for example, which are also tied to the machine.